### PR TITLE
[vividus-plugin-rest-api] Add step to save cookie as variable

### LINKED ISSUE #659: Required step to save api session cookie value to test variable
Required step to work with api session cookies: find cookie value by name and save to test variable.
Example:
```gherkin
When I save value of HTTP cookie with name `$cookieName` to $scopes variable `$variableName`
```
If there are cookies with the same names, but different additional parameters (path, isSecure, version, expiryDate etc), find first cookie ignoring additional parameters and save to test variable.

### PR DESCRIPTION
**Pull request by EDbarvinsky**
Fixes #659

### CODE REVIEW

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpCookieSteps.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+     * <i>Available scopes:</i>
+     * <ul>
+     * <li><b>STEP</b> - the variable will be available only within the step,
+     * <li><b>SCENARIO</b> - the variable will be available only within the scenario,
+     * <li><b>STORY</b> - the variable will be available within the whole story,
+     * <li><b>NEXT_BATCHES</b> - the variable will be available starting from next batch
+     * </ul>
+     * @param variableName name of variable
+     */
+    @When("I save value of HTTP cookie with name `$cookieName` to $scopes variable `$variableName`")
```
Comment (ikalinin1) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpCookieSteps.java:56]: I propose to use a dynamic variable in favor of step;
Just like for the body

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpCookieStepsTests.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    @Mock
+    private ISoftAssert softAssert;
+
+    @InjectMocks
+    private HttpCookieSteps httpCookieSteps;
+
+    @Test
+    void shouldSaveCookieValueIntoVariableContext()
+    {
+        CookieStore cookieStore = Mockito.mock(CookieStore.class);
```
Comment (AlenaBudevich) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpCookieStepsTests.java:64]: maybe make sense to move
```
CookieStore cookieStore = Mockito.mock(CookieStore.class);
when(cookieStoreProvider.getCookieStore()).thenReturn(cookieStore);
```
in @before ?

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpCookieSteps.java (hunk 1 above)
Comment (ikalinin1) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpCookieSteps.java:56]: Discussed with the team, the step is the simpler API due to cookies complexity

File: vividus-tests/src/main/resources/story/integration/RestApiStepsTests.story [hunk 3]
```diff
@@ -36,3 +36,8 @@ Scenario: Verify step "Then the connection is secured using $securityProtocol pr
 Given request body: ${request-body}
 When I send HTTP DELETE to the relative URL '/delete'
 Then the connection is secured using TLSv1.2 protocol
+
+Scenario: Verify HTTP cookie is saved to local storage
```
[the remark this answers is not in the document]
Comment (ikalinin1) [reply vividus-tests/src/main/resources/story/integration/RestApiStepsTests.story:40]: Do we follow a template Verify step: 'When I save value of HTTP cookie with name `$cookieName` to $scopes variable `$variableName`'

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpCookieSteps.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    public void saveHttpCookieIntoVariable(String cookieName, Set<VariableScope> scopes, String variableName)
+    {
+        List<Cookie> cookies = cookieStoreProvider.getCookieStore().getCookies()
+                .stream()
+                .filter(cookie -> cookie.getName().equals(cookieName))
+                .collect(Collectors.toList());
+        softAssert.assertTrue(String.format("Cookie with the name '%s' is set", cookieName), cookies.size() == 1);
```
Comment (dKatechev) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpCookieSteps.java:65]: If we have more than 1 cookies with the same name, will this step successfully pass?  
What is the suggested solution to handle such situation?

**Comment by EDbarvinsky** [comment]
Retry DeepCode

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpCookieSteps.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                    bddVariableContext.putVariable(scopes, variableName, filteredCookies.get(0).getValue());
+                }
+                else
+                {
+                    softAssert.recordFailedAssertion(
+                            String.format("Not found cookie with name '%s' and path '/' or was several candidates",
+                                    cookieName));
+                }
+            }
+        }
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpCookieSteps.java:95]:
```java
List<Cookie> cookies = cookieStoreProvider.getCookieStore().getCookies().stream()
                .filter(cookie -> cookie.getName().equals(cookieName))
                .collect(Collectors.toList());
int cookiesNumber = cookies.size();
if (softAssert.assertThat(String.format("Number of cookies with name '%s'", cookieName), cookiesNumber, greaterThan(0)))
{
    if (cookiesNumber == 1)
    {
        bddVariableContext.putVariable(scopes, variableName, cookies.get(0).getValue());
    }
    else
    {
        String rootPath = "/";
        LOGGER.info("Filtering cookies by path attribute '{}'", rootPath);
        cookies = cookies.stream()
                .filter(cookie -> rootPath.equals(cookie.getPath()))
                .collect(Collectors.toList());
        if (softAssert.assertEquals(String.format("Number of cookies with name '%s' and path attribute '%s'", cookieName, rootPath), 1, cookies.size())
        {
            bddVariableContext.putVariable(scopes, variableName, cookies.get(0).getValue());
        }
    }
}

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpCookieStepsTests.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        verifyCookieAssertion(cookies);
+        verifyNoInteractions(bddVariableContext);
+        verifyNoMoreInteractions(softAssert);
+    }
+
+    private void verifyCookieAssertion(List<Cookie> cookies)
+    {
+        verify(softAssert)
+                .assertTrue(eq(String.format(PRESENT_COOKIES_WITH_THE_NAME, cookies.size(), NAME)),
+                        eq(!cookies.isEmpty()));
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpCookieStepsTests.java:147]: formatting a bit awkward

**Pull request merged**